### ISSUE TITLE
kubelet should optionally skip writing hostname entry in /etc/hosts

### ISSUE THREAD

**Issue by mlowery** [kubernetes/kubernetes#57039]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:

We have a use case that requires pods to have custom DNS records (i.e. not the cluster's default domain). Additionally, `hostname -f` must return this custom FQDN. Summarizing these into technical attributes:

* Let's assume there exists a controller that manages the DNS records. It observes pods and creates the custom DNS records.
* Let's assume `spec.hostname` is set.
* Let's assume that the pods have a proper `search` in the injected `/etc/resolv.conf`, such that the custom FQDN can be resolved when pairing `spec.hostname` with one of the `search` entries.
* Finally, let's assume a kubelet-managed `/etc/hosts`.
* To make `hostname -f` work, kubelet must not write the line containing `spec.hostname` to `/etc/hosts` (since doing so prevents any nameserver-based resolution). Today, kubelet unconditionally writes such an entry to `/etc/hosts` (the last line). We need to allow kubelet to optionally skip the writing of this line which then allows a nameserver-based resolution to occur when `hostname -f` is run.

**What you expected to happen**:

To summarize the feature request, can kubelet be made to optionally not write the pod IP to `/etc/hosts`? Essentially, a user who enables this on a pod is saying, "I want a nameserver-based resolution to occur for this pod when resolving its own hostname."

**How to reproduce it (as minimally and precisely as possible)**:

It seems that a workaround is to delete this line for each container in the pod on each container start.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.6.3
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/sig network
@kubernetes/sig-network
